helix-dap/src/client.rs

```diff
@@ -0,0 +1,343 @@
+                                name
+                            ),
+                        },
+                        None => {
+                            info!("unhandled event {}", name);
+                            // client_tx.send(Payload::Event(ev)).expect("Failed to send");
+                        }
+                    }
+                }
+                Payload::Response(_) => unreachable!(),
+                Payload::Request(_) => todo!(),
+            }
+        }
+    }
+
+    pub async fn listen_for_event(&self, name: String) -> Receiver<Event> {
+        let (rx, tx) = channel(1);
+        self.awaited_events.lock().await.insert(name.clone(), rx);
+        tx
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    fn next_request_id(&self) -> u64 {
+        self.request_counter.fetch_add(1, Ordering::Relaxed)
+    }
+
+    async fn request<R: crate::types::Request>(
+        &self,
+        arguments: R::Arguments,
+    ) -> Result<R::Result> {
+        let (callback_tx, mut callback_rx) = channel(1);
+
+        let arguments = Some(serde_json::to_value(arguments)?);
+
+        let req = Request {
+            back_ch: Some(callback_tx),
+            seq: self.next_request_id(),
+            command: R::COMMAND.to_string(),
+            arguments,
+        };
+
+        self.server_tx
+            .send(req)
+            .expect("Failed to send request to debugger");
+
+        let response = callback_rx.recv().await.unwrap()?;
+        let response = serde_json::from_value(response.body.unwrap_or_default())?;
+        Ok(response)
+    }
+
+    pub fn capabilities(&self) -> &DebuggerCapabilities {
+        self.capabilities
+            .as_ref()
+            .expect("debugger not yet initialized!")
+    }
+
+    pub async fn initialize(&mut self, adapter_id: String) -> Result<()> {
+        let args = requests::InitializeArguments {
+            client_id: Some("hx".to_owned()),
+            client_name: Some("helix".to_owned()),
+            adapter_id,
+            locale: Some("en-us".to_owned()),
+            lines_start_at_one: Some(true),
+            columns_start_at_one: Some(true),
+            path_format: Some("path".to_owned()),
+            supports_variable_type: Some(false),
+            supports_variable_paging: Some(false),
+            supports_run_in_terminal_request: Some(false),
+            supports_memory_references: Some(false),
+            supports_progress_reporting: Some(false),
+            supports_invalidated_event: Some(false),
+        };
+
+        let response = self.request::<requests::Initialize>(args).await?;
+        self.capabilities = Some(response);
+
+        Ok(())
+    }
+
+    pub async fn disconnect(&mut self) -> Result<()> {
+        self.request::<requests::Disconnect>(()).await
+    }
+
+    pub async fn launch(&mut self, args: serde_json::Value) -> Result<()> {
+        // TODO: buffer these until initialized arrives
```

Comment:
`initialized` will get emitted after program is loaded for debugging, not before. Will remove these TODOs.

Reply:
`initialized` gets sent after the `initialize` request and **before** launch. It's invalid to send `launch` before this event.

https://microsoft.github.io/debug-adapter-protocol/specification#Events_Initialized

The sequence of events/requests is as follows:

> 
>     adapters sends ‘initialized’ event (after the ‘initialize’ request has returned)
>     frontend sends zero or more ‘setBreakpoints’ requests
>     frontend sends one ‘setFunctionBreakpoints’ request (if capability ‘supportsFunctionBreakpoints’ is true)
>     frontend sends a ‘setExceptionBreakpoints’ request if one or more ‘exceptionBreakpointFilters’ have been defined (or if ‘supportsConfigurationDoneRequest’ is not defined or false)
>     frontend sends other future configuration requests
>     frontend sends one ‘configurationDone’ request to indicate the end of the configuration.


`dap-mode` waits for initialized event, then sends breakpoints & configurationDone: https://github.com/emacs-lsp/dap-mode/blob/8ec7e98986ea46e3c36b7ecbdf9a6562c90c7632/dap-mode.el#L956-L960